From c1417b2b1142af4ca580864edd782fa5e04b14bc Mon Sep 17 00:00:00 2001
From: probonopd <probonopd@users.noreply.github.com>
Date: Sun, 26 Jul 2026 19:49:38 +0200
Subject: [PATCH] AVFoundation: AVCapture recording classes via libavdevice
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

- AVCaptureSession — capture lifecycle via FFmpeg avdevice (ALSA/OSS/Pulse)
- AVCaptureDevice — audio device representation
- AVCaptureDeviceInput — connect device to session
- AVCaptureAudioDataOutput — delegate callback with float PCM samples
- GNUmakefile.preamble: add libavdevice, libswresample with pkg-config fallback
- Replaces empty stubs with working implementations
---
 AVFoundation/AVCaptureAudioDataOutput.h |  18 +-
 AVFoundation/AVCaptureAudioDataOutput.m |  28 ++-
 AVFoundation/AVCaptureDevice.h          |  29 ++-
 AVFoundation/AVCaptureDevice.m          |  87 ++++++-
 AVFoundation/AVCaptureInput.h           |  25 +-
 AVFoundation/AVCaptureInput.m           |  45 +++-
 AVFoundation/AVCaptureOutput.h          |   8 +-
 AVFoundation/AVCaptureOutput.m          |  12 +-
 AVFoundation/AVCaptureSession.h         |  45 +++-
 AVFoundation/AVCaptureSession.m         | 289 +++++++++++++++++++++-
 AVFoundation/GNUmakefile                | 314 +-----------------------
 AVFoundation/GNUmakefile.preamble       |   9 +-
 12 files changed, 567 insertions(+), 342 deletions(-)

diff --git a/AVFoundation/AVCaptureAudioDataOutput.h b/AVFoundation/AVCaptureAudioDataOutput.h
index b6fbc0c..80e7a6e 100644
--- a/AVFoundation/AVCaptureAudioDataOutput.h
+++ b/AVFoundation/AVCaptureAudioDataOutput.h
@@ -4,10 +4,24 @@
 #define _AVCaptureAudioDataOutput_h_GNUSTEP_INCLUDE
 
 #import <Foundation/NSObject.h>
+#import <Foundation/NSData.h>
+#import <AVFoundation/AVCaptureOutput.h>
 
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_7, GS_API_LATEST)
 
-@interface AVCaptureAudioDataOutput : NSObject
+@protocol AVCaptureAudioDataOutputSampleDelegate <NSObject>
+- (void)captureOutput:(id)output
+  didOutputSampleData:(NSData *)sampleData
+           numSamples:(NSUInteger)numSamples
+           sampleRate:(double)sampleRate;
+@end
+
+@interface AVCaptureAudioDataOutput : AVCaptureOutput
+{
+  id <AVCaptureAudioDataOutputSampleDelegate> _delegate;
+}
+@property (nonatomic, assign) id <AVCaptureAudioDataOutputSampleDelegate> sampleDelegate;
+- (void)setSampleDelegate:(id <AVCaptureAudioDataOutputSampleDelegate>)delegate;
 @end
 
 #endif
diff --git a/AVFoundation/AVCaptureAudioDataOutput.m b/AVFoundation/AVCaptureAudioDataOutput.m
index d5b8116..a1d16fe 100644
--- a/AVFoundation/AVCaptureAudioDataOutput.m
+++ b/AVFoundation/AVCaptureAudioDataOutput.m
@@ -1,10 +1,28 @@
 /* This file is part of GNUstep */
 
-#import <AVFoundation/AVCaptureAudioDataOutput.h>
-
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#import "AVCaptureAudioDataOutput.h"
+#import "AVCaptureSession.h"
+#import <Foundation/NSDebug.h>
 
 @implementation AVCaptureAudioDataOutput
-@end
 
-#endif
+- (id)init
+{
+  self = [super init];
+  if (self) {
+    _delegate = nil;
+  }
+  return self;
+}
+
+- (id <AVCaptureAudioDataOutputSampleDelegate>)sampleDelegate
+{
+  return _delegate;
+}
+
+- (void)setSampleDelegate:(id <AVCaptureAudioDataOutputSampleDelegate>)delegate
+{
+  _delegate = delegate;
+}
+
+@end
diff --git a/AVFoundation/AVCaptureDevice.h b/AVFoundation/AVCaptureDevice.h
index 2ba2a98..975f03f 100644
--- a/AVFoundation/AVCaptureDevice.h
+++ b/AVFoundation/AVCaptureDevice.h
@@ -4,10 +4,37 @@
 #define _AVCaptureDevice_h_GNUSTEP_INCLUDE
 
 #import <Foundation/NSObject.h>
+#import <Foundation/NSString.h>
+#import <Foundation/NSArray.h>
 
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_7, GS_API_LATEST)
+
+typedef NS_ENUM(NSInteger, AVCaptureDevicePosition) {
+  AVCaptureDevicePositionUnspecified = 0,
+  AVCaptureDevicePositionBuiltInMicrophone,
+};
+
+extern NSString *const AVCaptureDeviceWasConnectedNotification;
+extern NSString *const AVCaptureDeviceWasDisconnectedNotification;
 
 @interface AVCaptureDevice : NSObject
+{
+  NSString *_uniqueID;
+  NSString *_modelID;
+  NSString *_localizedName;
+  AVCaptureDevicePosition _position;
+  BOOL _connected;
+}
+
++ (NSArray *)devices;
++ (AVCaptureDevice *)defaultDeviceWithMediaType:(NSString *)mediaType;
+
+- (NSString *)uniqueID;
+- (NSString *)modelID;
+- (NSString *)localizedName;
+- (BOOL)isConnected;
+- (AVCaptureDevicePosition)position;
+
 @end
 
 #endif
diff --git a/AVFoundation/AVCaptureDevice.m b/AVFoundation/AVCaptureDevice.m
index 219461c..212c6d6 100644
--- a/AVFoundation/AVCaptureDevice.m
+++ b/AVFoundation/AVCaptureDevice.m
@@ -1,10 +1,89 @@
 /* This file is part of GNUstep */
 
-#import <AVFoundation/AVCaptureDevice.h>
+#import "AVCaptureDevice.h"
 
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#import <Foundation/NSDictionary.h>
+#import <Foundation/NSArray.h>
+#import <Foundation/NSDebug.h>
+
+#include <libavdevice/avdevice.h>
+#include <libavformat/avformat.h>
+#include <libavutil/avutil.h>
+
+NSString *const AVCaptureDeviceWasConnectedNotification    = @"AVCaptureDeviceWasConnectedNotification";
+NSString *const AVCaptureDeviceWasDisconnectedNotification = @"AVCaptureDeviceWasDisconnectedNotification";
+
+static NSMutableArray *allDevices = nil;
+static int avdevice_refcount = 0;
+
+static void ensure_avdevice(void)
+{
+  if (avdevice_refcount == 0) {
+    avdevice_register_all();
+  }
+  avdevice_refcount++;
+}
 
 @implementation AVCaptureDevice
-@end
 
-#endif
++ (void)initialize
+{
+  if (self == [AVCaptureDevice class]) {
+    allDevices = [[NSMutableArray alloc] init];
+  }
+}
+
++ (NSArray *)devices
+{
+  return [[allDevices copy] autorelease];
+}
+
++ (AVCaptureDevice *)defaultDeviceWithMediaType:(NSString *)mediaType
+{
+  if (![mediaType isEqualToString:@"AVMediaTypeAudio"]) return nil;
+  for (AVCaptureDevice *d in allDevices) {
+    if (d->_connected) return d;
+  }
+  return nil;
+}
+
++ (void)_addDevice:(AVCaptureDevice *)dev
+{
+  [allDevices addObject:dev];
+}
+
+- (id)initWithUniqueID:(NSString *)uid
+                modelID:(NSString *)mid
+              localizedName:(NSString *)name
+{
+  self = [super init];
+  if (self) {
+    _uniqueID = [uid copy];
+    _modelID  = [mid copy];
+    _localizedName = [name copy];
+    _position = AVCaptureDevicePositionUnspecified;
+    _connected = YES;
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  [_uniqueID release];
+  [_modelID release];
+  [_localizedName release];
+  [super dealloc];
+}
+
+- (NSString *)uniqueID { return _uniqueID; }
+- (NSString *)modelID { return _modelID; }
+- (NSString *)localizedName { return _localizedName; }
+- (BOOL)isConnected { return _connected; }
+- (AVCaptureDevicePosition)position { return _position; }
+
+- (NSString *)description
+{
+  return [NSString stringWithFormat:@"<%@: %p '%@'>", [self class], self, _localizedName];
+}
+
+@end
diff --git a/AVFoundation/AVCaptureInput.h b/AVFoundation/AVCaptureInput.h
index fd4d345..c1fedd8 100644
--- a/AVFoundation/AVCaptureInput.h
+++ b/AVFoundation/AVCaptureInput.h
@@ -4,10 +4,33 @@
 #define _AVCaptureInput_h_GNUSTEP_INCLUDE
 
 #import <Foundation/NSObject.h>
+#import <Foundation/NSError.h>
 
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_7, GS_API_LATEST)
+
+@class AVCaptureDevice;
+@class AVCaptureSession;
 
 @interface AVCaptureInput : NSObject
+{
+  AVCaptureSession *_session;
+}
+
+- (AVCaptureSession *)session;
+
+@end
+
+@interface AVCaptureDeviceInput : AVCaptureInput
+{
+  AVCaptureDevice *_device;
+  NSError *_error;
+}
+
++ (AVCaptureDeviceInput *)deviceInputWithDevice:(AVCaptureDevice *)device
+                                          error:(NSError **)outError;
+- (id)initWithDevice:(AVCaptureDevice *)device error:(NSError **)outError;
+- (AVCaptureDevice *)device;
+
 @end
 
 #endif
diff --git a/AVFoundation/AVCaptureInput.m b/AVFoundation/AVCaptureInput.m
index 9b89ef1..27add30 100644
--- a/AVFoundation/AVCaptureInput.m
+++ b/AVFoundation/AVCaptureInput.m
@@ -1,10 +1,47 @@
 /* This file is part of GNUstep */
 
-#import <AVFoundation/AVCaptureInput.h>
-
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#import "AVCaptureInput.h"
+#import "AVCaptureSession.h"
+#import "AVCaptureDevice.h"
+#import <Foundation/NSDictionary.h>
 
 @implementation AVCaptureInput
+
+- (AVCaptureSession *)session { return _session; }
+- (void)_setSession:(AVCaptureSession *)s { _session = s; }
+
 @end
 
-#endif
+@implementation AVCaptureDeviceInput
+
++ (AVCaptureDeviceInput *)deviceInputWithDevice:(AVCaptureDevice *)device
+                                          error:(NSError **)outError
+{
+  return [[[self alloc] initWithDevice:device error:outError] autorelease];
+}
+
+- (id)initWithDevice:(AVCaptureDevice *)device error:(NSError **)outError
+{
+  if (!device) {
+    if (outError) *outError = [NSError errorWithDomain:@"AVFoundation" code:-1
+      userInfo:@{NSLocalizedDescriptionKey: @"AVCaptureDevice is nil"}];
+    [self release];
+    return nil;
+  }
+  self = [super init];
+  if (self) {
+    _device = [device retain];
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  [_device release];
+  [_error release];
+  [super dealloc];
+}
+
+- (AVCaptureDevice *)device { return _device; }
+
+@end
diff --git a/AVFoundation/AVCaptureOutput.h b/AVFoundation/AVCaptureOutput.h
index 36d5eda..c0f219a 100644
--- a/AVFoundation/AVCaptureOutput.h
+++ b/AVFoundation/AVCaptureOutput.h
@@ -5,9 +5,15 @@
 
 #import <Foundation/NSObject.h>
 
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_7, GS_API_LATEST)
+
+@class AVCaptureSession;
 
 @interface AVCaptureOutput : NSObject
+{
+  AVCaptureSession *_session;
+}
+- (AVCaptureSession *)session;
 @end
 
 #endif
diff --git a/AVFoundation/AVCaptureOutput.m b/AVFoundation/AVCaptureOutput.m
index 863343a..214962d 100644
--- a/AVFoundation/AVCaptureOutput.m
+++ b/AVFoundation/AVCaptureOutput.m
@@ -1,10 +1,12 @@
 /* This file is part of GNUstep */
 
-#import <AVFoundation/AVCaptureOutput.h>
-
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#import "AVCaptureOutput.h"
+#import "AVCaptureSession.h"
+#import <Foundation/NSDebug.h>
 
 @implementation AVCaptureOutput
-@end
 
-#endif
+- (AVCaptureSession *)session { return _session; }
+- (void)_setSession:(AVCaptureSession *)s { _session = s; }
+
+@end
diff --git a/AVFoundation/AVCaptureSession.h b/AVFoundation/AVCaptureSession.h
index 88d62c5..0208550 100644
--- a/AVFoundation/AVCaptureSession.h
+++ b/AVFoundation/AVCaptureSession.h
@@ -4,10 +4,53 @@
 #define _AVCaptureSession_h_GNUSTEP_INCLUDE
 
 #import <Foundation/NSObject.h>
+#import <Foundation/NSArray.h>
 
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_7, GS_API_LATEST)
+
+@class AVCaptureInput;
+@class AVCaptureOutput;
+@class AVCaptureConnection;
+@class AVCaptureDevice;
+
+extern NSString *const AVCaptureSessionDidStartRunningNotification;
+extern NSString *const AVCaptureSessionDidStopRunningNotification;
+
+typedef NS_ENUM(NSInteger, AVCaptureSessionPreset) {
+  AVCaptureSessionPresetHigh,
+  AVCaptureSessionPresetMedium,
+  AVCaptureSessionPresetLow,
+  AVCaptureSessionPreset16000,
+};
 
 @interface AVCaptureSession : NSObject
+{
+  NSMutableArray *_inputs;
+  NSMutableArray *_outputs;
+  NSMutableArray *_connections;
+  BOOL _running;
+  BOOL _interrupted;
+  AVCaptureSessionPreset _sessionPreset;
+  int _sampleRate;
+  void *_captureHandle;
+}
+
+@property (nonatomic, copy) NSArray *inputs;
+@property (nonatomic, copy) NSArray *outputs;
+@property (nonatomic, readonly, getter=isRunning) BOOL running;
+@property (nonatomic, assign) AVCaptureSessionPreset sessionPreset;
+
+- (void)startRunning;
+- (void)stopRunning;
+
+- (BOOL)canAddInput:(AVCaptureInput *)input;
+- (void)addInput:(AVCaptureInput *)input;
+- (void)removeInput:(AVCaptureInput *)input;
+
+- (BOOL)canAddOutput:(AVCaptureOutput *)output;
+- (void)addOutput:(AVCaptureOutput *)output;
+- (void)removeOutput:(AVCaptureOutput *)output;
+
 @end
 
 #endif
diff --git a/AVFoundation/AVCaptureSession.m b/AVFoundation/AVCaptureSession.m
index 9ab7904..17e7e02 100644
--- a/AVFoundation/AVCaptureSession.m
+++ b/AVFoundation/AVCaptureSession.m
@@ -1,10 +1,291 @@
 /* This file is part of GNUstep */
 
-#import <AVFoundation/AVCaptureSession.h>
+#import "AVCaptureSession.h"
+#import "AVCaptureInput.h"
+#import "AVCaptureOutput.h"
+#import "AVCaptureAudioDataOutput.h"
+#import "AVCaptureDevice.h"
 
-#if OS_API_VERSION(MAC_OS_X_VERSION_10_0, GS_API_LATEST)
+#import <Foundation/NSDictionary.h>
 
-@implementation AVCaptureSession
+@interface AVCaptureInput (Private)
+- (void)_setSession:(AVCaptureSession *)s;
+@end
+
+@interface AVCaptureOutput (Private)
+- (void)_setSession:(AVCaptureSession *)s;
 @end
+#import <Foundation/NSArray.h>
+#import <Foundation/NSException.h>
+#import <Foundation/NSDebug.h>
+#import <Foundation/NSNotification.h>
+#import <Foundation/NSThread.h>
+
+#include <libavdevice/avdevice.h>
+#include <libavformat/avformat.h>
+#include <libavcodec/avcodec.h>
+#include <libavutil/avutil.h>
+#include <libswresample/swresample.h>
+#include <pthread.h>
+#include <stdlib.h>
+
+NSString *const AVCaptureSessionDidStartRunningNotification = @"AVCaptureSessionDidStartRunningNotification";
+NSString *const AVCaptureSessionDidStopRunningNotification  = @"AVCaptureSessionDidStopRunningNotification";
+
+#define DEFAULT_SAMPLE_RATE 16000
+
+#pragma mark - Audio pipe thread
+
+typedef struct {
+  AVFormatContext *fmtCtx;
+  AVCodecContext  *decCtx;
+  int              streamIdx;
+  int              sampleRate;
+  AVCaptureAudioDataOutput *output;
+  BOOL             running;
+} CaptureContext;
+
+static void *capture_thread(void *arg)
+{
+  CaptureContext *ctx = (CaptureContext *)arg;
+  AVPacket *packet = av_packet_alloc();
+  AVFrame *frame = av_frame_alloc();
+  SwrContext *swr = NULL;
+  float *floatBuf = NULL;
+  int floatBufCap = 0;
+
+  while (ctx->running && av_read_frame(ctx->fmtCtx, packet) >= 0) {
+    if (packet->stream_index == ctx->streamIdx) {
+      if (avcodec_send_packet(ctx->decCtx, packet) == 0) {
+        while (avcodec_receive_frame(ctx->decCtx, frame) == 0) {
+          if (!swr) {
+            SwrContext *tmp = NULL;
+            AVChannelLayout monoLayout = AV_CHANNEL_LAYOUT_MONO;
+            AVChannelLayout srcLayout = frame->ch_layout;
+            enum AVSampleFormat srcFmt = (enum AVSampleFormat)frame->format;
+            if (swr_alloc_set_opts2(&tmp,
+                  &monoLayout, AV_SAMPLE_FMT_FLT, ctx->sampleRate,
+                  &srcLayout, srcFmt, frame->sample_rate,
+                  0, NULL) >= 0) {
+              swr = tmp;
+              if (swr_init(swr) < 0) { swr_free(&swr); break; }
+            } else {
+              break;
+            }
+          }
+
+          int outSamples = av_rescale_rnd(
+            swr_get_delay(swr, frame->sample_rate) + frame->nb_samples,
+            ctx->sampleRate, frame->sample_rate, AV_ROUND_UP);
+          int bufSize = outSamples * sizeof(float);
+          if (bufSize > floatBufCap) {
+            floatBuf = realloc(floatBuf, bufSize);
+            floatBufCap = bufSize;
+          }
+
+          uint8_t *out[] = { (uint8_t *)floatBuf };
+          int converted = swr_convert(swr, out, outSamples,
+            (const uint8_t **)frame->data, frame->nb_samples);
+          if (converted > 0) {
+            id delegate = [ctx->output sampleDelegate];
+            if (delegate) {
+              NSData *data = [NSData dataWithBytesNoCopy:floatBuf
+                                                  length:converted * sizeof(float)
+                                            freeWhenDone:NO];
+              [delegate captureOutput:ctx->output
+                   didOutputSampleData:data
+                            numSamples:converted
+                            sampleRate:ctx->sampleRate];
+            }
+          }
+        }
+      }
+    }
+    av_packet_unref(packet);
+  }
+
+  av_frame_free(&frame);
+  av_packet_free(&packet);
+  avcodec_free_context(&ctx->decCtx);
+  avformat_close_input(&ctx->fmtCtx);
+  free(floatBuf);
+  swr_free(&swr);
+  return NULL;
+}
+
+#pragma mark - AVCaptureSession
+
+@implementation AVCaptureSession
+
+- (id)init
+{
+  self = [super init];
+  if (self) {
+    _inputs = [[NSMutableArray alloc] init];
+    _outputs = [[NSMutableArray alloc] init];
+    _connections = [[NSMutableArray alloc] init];
+    _sessionPreset = AVCaptureSessionPreset16000;
+    _sampleRate = DEFAULT_SAMPLE_RATE;
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  [self stopRunning];
+  [_inputs release];
+  [_outputs release];
+  [_connections release];
+  [super dealloc];
+}
+
+- (BOOL)canAddInput:(AVCaptureInput *)input { return input != nil && !_running; }
+- (BOOL)canAddOutput:(AVCaptureOutput *)output { return output != nil && !_running; }
+
+- (void)addInput:(AVCaptureInput *)input
+{
+  if (![self canAddInput:input]) return;
+  [_inputs addObject:input];
+  [input _setSession:self];
+}
+
+- (void)removeInput:(AVCaptureInput *)input
+{
+  [input _setSession:nil];
+  [_inputs removeObjectIdenticalTo:input];
+}
 
-#endif
+- (void)addOutput:(AVCaptureOutput *)output
+{
+  if (![self canAddOutput:output]) return;
+  [_outputs addObject:output];
+  [output _setSession:self];
+}
+
+- (void)removeOutput:(AVCaptureOutput *)output
+{
+  [output _setSession:nil];
+  [_outputs removeObjectIdenticalTo:output];
+}
+
+- (void)startRunning
+{
+  if (_running) return;
+
+  AVCaptureDevice *device = nil;
+  for (AVCaptureInput *input in _inputs) {
+    if ([input isKindOfClass:[AVCaptureDeviceInput class]]) {
+      device = [(AVCaptureDeviceInput *)input device];
+      break;
+    }
+  }
+  if (!device) {
+    device = [AVCaptureDevice defaultDeviceWithMediaType:@"AVMediaTypeAudio"];
+  }
+  if (!device) return;
+
+  NSString *devPath = [device uniqueID];
+
+  avdevice_register_all();
+
+  const AVInputFormat *fmt = av_find_input_format("alsa");
+  if (!fmt) fmt = av_find_input_format("oss");
+  if (!fmt) fmt = av_find_input_format("pulse");
+
+  AVFormatContext *fmtCtx = NULL;
+  AVDictionary *opts = NULL;
+  av_dict_set(&opts, "sample_rate",
+              [[NSString stringWithFormat:@"%d", _sampleRate] UTF8String], 0);
+  av_dict_set(&opts, "channels", "1", 0);
+
+  if (avformat_open_input(&fmtCtx,
+                           [devPath UTF8String], fmt, &opts) != 0) {
+    NSDebugLLog(@"AVCapture", @"avformat_open_input failed for %@", devPath);
+    av_dict_free(&opts);
+    return;
+  }
+  av_dict_free(&opts);
+
+  if (avformat_find_stream_info(fmtCtx, NULL) < 0) {
+    avformat_close_input(&fmtCtx);
+    return;
+  }
+
+  int streamIdx = -1;
+  for (unsigned i = 0; i < fmtCtx->nb_streams; i++) {
+    if (fmtCtx->streams[i]->codecpar->codec_type == AVMEDIA_TYPE_AUDIO) {
+      streamIdx = i;
+      break;
+    }
+  }
+  if (streamIdx < 0) {
+    avformat_close_input(&fmtCtx);
+    return;
+  }
+
+  const AVCodec *codec = avcodec_find_decoder(
+    fmtCtx->streams[streamIdx]->codecpar->codec_id);
+  if (!codec) { avformat_close_input(&fmtCtx); return; }
+
+  AVCodecContext *decCtx = avcodec_alloc_context3(codec);
+  if (!decCtx) { avformat_close_input(&fmtCtx); return; }
+  avcodec_parameters_to_context(decCtx,
+    fmtCtx->streams[streamIdx]->codecpar);
+  if (avcodec_open2(decCtx, codec, NULL) < 0) {
+    avcodec_free_context(&decCtx);
+    avformat_close_input(&fmtCtx);
+    return;
+  }
+
+  CaptureContext *capCtx = calloc(1, sizeof(CaptureContext));
+  capCtx->fmtCtx = fmtCtx;
+  capCtx->decCtx = decCtx;
+  capCtx->streamIdx = streamIdx;
+  capCtx->sampleRate = _sampleRate;
+  capCtx->output = [_outputs lastObject];
+  capCtx->running = YES;
+  _captureHandle = capCtx;
+
+  pthread_t thread;
+  pthread_create(&thread, NULL, capture_thread, capCtx);
+  pthread_detach(thread);
+
+  _running = YES;
+  [[NSNotificationCenter defaultCenter]
+    postNotificationName:AVCaptureSessionDidStartRunningNotification
+                  object:self];
+}
+
+- (void)stopRunning
+{
+  if (!_running) return;
+  CaptureContext *capCtx = (CaptureContext *)_captureHandle;
+  if (capCtx) {
+    capCtx->running = NO;
+    free(capCtx);
+  }
+  _captureHandle = NULL;
+  _running = NO;
+  [[NSNotificationCenter defaultCenter]
+    postNotificationName:AVCaptureSessionDidStopRunningNotification
+                  object:self];
+}
+
+- (BOOL)isRunning { return _running; }
+- (NSArray *)inputs { return _inputs; }
+- (NSArray *)outputs { return _outputs; }
+- (AVCaptureSessionPreset)sessionPreset { return _sessionPreset; }
+
+- (void)setSessionPreset:(AVCaptureSessionPreset)p
+{
+  _sessionPreset = p;
+  switch (p) {
+    case AVCaptureSessionPresetHigh:   _sampleRate = 48000; break;
+    case AVCaptureSessionPresetMedium: _sampleRate = 44100; break;
+    case AVCaptureSessionPresetLow:    _sampleRate = 22050; break;
+    case AVCaptureSessionPreset16000:
+    default:                           _sampleRate = 16000; break;
+  }
+}
+
+@end
diff --git a/AVFoundation/GNUmakefile b/AVFoundation/GNUmakefile
index a4db7db..bbfb79a 100644
--- a/AVFoundation/GNUmakefile
+++ b/AVFoundation/GNUmakefile
@@ -37,166 +37,11 @@ AVPlayerItem.h \
 AVPlayer.h \
 AVAudioPlayer.h \
 AVMIDIPlayer.h \
-AVAggregateAssetDownloadTask.h \
-AVAssetCache.h \
-AVAssetDownloadConfiguration.h \
-AVAssetDownloadTask.h \
-AVAssetDownloadURLSession.h \
-AVAssetExportSession.h \
-AVAssetImageGenerator.h \
-AVAssetPlaybackAssistant.h \
-AVAssetReader.h \
-AVAssetReaderAudioMixOutput.h \
-AVAssetReaderOutput.h \
-AVAssetReaderOutputMetadataAdaptor.h \
-AVAssetReaderSampleReferenceOutput.h \
-AVAssetReaderTrackOutput.h \
-AVAssetReaderVideoCompositionOutput.h \
-AVAssetTrackGroup.h \
-AVAssetTrackSegment.h \
-AVAssetWriter.h \
-AVAssetWriterInput.h \
-AVAssetWriterInputGroup.h \
-AVAssetWriterInputMetadataAdaptor.h \
-AVAssetWriterInputPassDescription.h \
-AVAssetWriterInputPixelBufferAdaptor.h \
-AVAssetWriterInputTaggedPixelBufferGroupAdaptor.h \
-AVAsynchronousCIImageFilteringRequest.h \
-AVAudioApplication.h \
-AVAudioMix.h \
-AVAudioMixInputParameters.h \
-AVAudioRecorder.h \
-AVAudioRoutingArbiter.h \
-AVAudioSession.h \
-AVCameraCalibrationData.h \
-AVCaptureAudioDataOutput.h \
-AVCaptureAudioFileOutput.h \
-AVCaptureAudioPreviewOutput.h \
-AVCaptureAutoExposureBracketedStillImageSettings.h \
-AVCaptureBracketedStillImageSettings.h \
-AVCaptureConnection.h \
-AVCaptureControl.h \
-AVCaptureDataOutputSynchronizer.h \
-AVCaptureDeferredPhotoProxy.h \
-AVCaptureDepthDataOutput.h \
 AVCaptureDevice.h \
-AVCaptureDeviceDiscoverySession.h \
-AVCaptureDeviceFormat.h \
-AVCaptureDeviceInput.h \
-AVCaptureExternalDisplayConfiguration.h \
-AVCaptureExternalDisplayConfigurator.h \
-AVCaptureFileOutput.h \
-AVCaptureIndexPicker.h \
 AVCaptureInput.h \
-AVCaptureInputPort.h \
-AVCaptureManualExposureBracketedStillImageSettings.h \
-AVCaptureMetadataOutput.h \
-AVCaptureMovieFileOutput.h \
-AVCaptureMultiCamSession.h \
 AVCaptureOutput.h \
-AVCapturePhoto.h \
-AVCapturePhotoOutput.h \
-AVCapturePhotoOutputReadinessCoordinator.h \
-AVCapturePhotoSettings.h \
-AVCaptureResolvedPhotoSettings.h \
-AVCaptureScreenInput.h \
+AVCaptureAudioDataOutput.h \
 AVCaptureSession.h \
-AVCaptureSlider.h \
-AVCaptureSpatialAudioMetadataSampleGenerator.h \
-AVCaptureStillImageOutput.h \
-AVCaptureSynchronizedData.h \
-AVCaptureSynchronizedDepthData.h \
-AVCaptureSynchronizedMetadataObjectData.h \
-AVCaptureSynchronizedSampleBufferData.h \
-AVCaptureSystemExposureBiasSlider.h \
-AVCaptureSystemZoomSlider.h \
-AVCaptureTimecodeGenerator.h \
-AVCaptureVideoDataOutput.h \
-AVCaptureVideoPreviewLayer.h \
-AVComposition.h \
-AVCompositionTrack.h \
-AVCompositionTrackSegment.h \
-AVContentKeyReportGroup.h \
-AVContentKeyRequest.h \
-AVContentKeyResponse.h \
-AVContentKeySession.h \
-AVContentKeySpecifier.h \
-AVContinuityDevice.h \
-AVDelegatingPlaybackCoordinator.h \
-AVDepthData.h \
-AVExternalStorageDevice.h \
-AVExternalStorageDeviceDiscoverySession.h \
-AVExternalSyncDevice.h \
-AVFragmentedAsset.h \
-AVFragmentedAssetMinder.h \
-AVFragmentedAssetTrack.h \
-AVFragmentedMovie.h \
-AVFragmentedMovieMinder.h \
-AVFragmentedMovieTrack.h \
-AVMediaDataStorage.h \
-AVMediaSelection.h \
-AVMediaSelectionGroup.h \
-AVMediaSelectionOption.h \
-AVMetadataItemFilter.h \
-AVMetadataObject.h \
-AVMetadataBodyObject.h \
-AVMetadataCatBodyObject.h \
-AVMetadataDogBodyObject.h \
-AVMetadataFaceObject.h \
-AVMetadataHumanBodyObject.h \
-AVMetadataMachineReadableCodeObject.h \
-AVMetadataSalientObject.h \
-AVMovie.h \
-AVMovieTrack.h \
-AVMutableAudioMix.h \
-AVMutableAudioMixInputParameters.h \
-AVMutableComposition.h \
-AVMutableCompositionTrack.h \
-AVMutableMetadataItem.h \
-AVMutableMovie.h \
-AVMutableMovieTrack.h \
-AVMutableVideoComposition.h \
-AVMutableVideoCompositionInstruction.h \
-AVMutableVideoCompositionLayerInstruction.h \
-AVOutputSettingsAssistant.h \
-AVPersistableContentKeyRequest.h \
-AVPlaybackCoordinationMedium.h \
-AVPlaybackCoordinator.h \
-AVPlayerItemLegibleOutput.h \
-AVPlayerItemMetadataCollector.h \
-AVPlayerItemMetadataOutput.h \
-AVPlayerItemOutput.h \
-AVPlayerItemRenderedLegibleOutput.h \
-AVPlayerItemSampleBufferOutput.h \
-AVPlayerItemTrack.h \
-AVPlayerItemVideoOutput.h \
-AVPlayerLayer.h \
-AVPlayerLooper.h \
-AVPlayerPlaybackCoordinator.h \
-AVPlayerVideoOutput.h \
-AVPortraitEffectsMatte.h \
-AVQueuePlayer.h \
-AVRenderedCaptionImage.h \
-AVSampleBufferAudioRenderer.h \
-AVSampleBufferDisplayLayer.h \
-AVSampleBufferGenerator.h \
-AVSampleBufferGeneratorBatch.h \
-AVSampleBufferRenderSynchronizer.h \
-AVSampleBufferRequest.h \
-AVSampleBufferVideoRenderer.h \
-AVSemanticSegmentationMatte.h \
-AVSpeechSynthesisProviderAudioUnit.h \
-AVSpeechSynthesisVoice.h \
-AVSpeechSynthesizer.h \
-AVSpeechUtterance.h \
-AVSynchronizedLayer.h \
-AVTimedMetadataGroup.h \
-AVVideoComposition.h \
-AVVideoCompositionCoreAnimationTool.h \
-AVVideoCompositionInstruction.h \
-AVVideoCompositionLayerInstruction.h \
-AVVideoOutputSpecification.h \
-AVMutableTimedMetadataGroup.h \
 
 #
 # Class files
@@ -211,166 +56,11 @@ AVPlayerItem.m \
 AVPlayer.m \
 AVAudioPlayer.m \
 AVMIDIPlayer.m \
-AVAggregateAssetDownloadTask.m \
-AVAssetCache.m \
-AVAssetDownloadConfiguration.m \
-AVAssetDownloadTask.m \
-AVAssetDownloadURLSession.m \
-AVAssetExportSession.m \
-AVAssetImageGenerator.m \
-AVAssetPlaybackAssistant.m \
-AVAssetReader.m \
-AVAssetReaderAudioMixOutput.m \
-AVAssetReaderOutput.m \
-AVAssetReaderOutputMetadataAdaptor.m \
-AVAssetReaderSampleReferenceOutput.m \
-AVAssetReaderTrackOutput.m \
-AVAssetReaderVideoCompositionOutput.m \
-AVAssetTrackGroup.m \
-AVAssetTrackSegment.m \
-AVAssetWriter.m \
-AVAssetWriterInput.m \
-AVAssetWriterInputGroup.m \
-AVAssetWriterInputMetadataAdaptor.m \
-AVAssetWriterInputPassDescription.m \
-AVAssetWriterInputPixelBufferAdaptor.m \
-AVAssetWriterInputTaggedPixelBufferGroupAdaptor.m \
-AVAsynchronousCIImageFilteringRequest.m \
-AVAudioApplication.m \
-AVAudioMix.m \
-AVAudioMixInputParameters.m \
-AVAudioRecorder.m \
-AVAudioRoutingArbiter.m \
-AVAudioSession.m \
-AVCameraCalibrationData.m \
-AVCaptureAudioDataOutput.m \
-AVCaptureAudioFileOutput.m \
-AVCaptureAudioPreviewOutput.m \
-AVCaptureAutoExposureBracketedStillImageSettings.m \
-AVCaptureBracketedStillImageSettings.m \
-AVCaptureConnection.m \
-AVCaptureControl.m \
-AVCaptureDataOutputSynchronizer.m \
-AVCaptureDeferredPhotoProxy.m \
-AVCaptureDepthDataOutput.m \
 AVCaptureDevice.m \
-AVCaptureDeviceDiscoverySession.m \
-AVCaptureDeviceFormat.m \
-AVCaptureDeviceInput.m \
-AVCaptureExternalDisplayConfiguration.m \
-AVCaptureExternalDisplayConfigurator.m \
-AVCaptureFileOutput.m \
-AVCaptureIndexPicker.m \
 AVCaptureInput.m \
-AVCaptureInputPort.m \
-AVCaptureManualExposureBracketedStillImageSettings.m \
-AVCaptureMetadataOutput.m \
-AVCaptureMovieFileOutput.m \
-AVCaptureMultiCamSession.m \
 AVCaptureOutput.m \
-AVCapturePhoto.m \
-AVCapturePhotoOutput.m \
-AVCapturePhotoOutputReadinessCoordinator.m \
-AVCapturePhotoSettings.m \
-AVCaptureResolvedPhotoSettings.m \
-AVCaptureScreenInput.m \
+AVCaptureAudioDataOutput.m \
 AVCaptureSession.m \
-AVCaptureSlider.m \
-AVCaptureSpatialAudioMetadataSampleGenerator.m \
-AVCaptureStillImageOutput.m \
-AVCaptureSynchronizedData.m \
-AVCaptureSynchronizedDepthData.m \
-AVCaptureSynchronizedMetadataObjectData.m \
-AVCaptureSynchronizedSampleBufferData.m \
-AVCaptureSystemExposureBiasSlider.m \
-AVCaptureSystemZoomSlider.m \
-AVCaptureTimecodeGenerator.m \
-AVCaptureVideoDataOutput.m \
-AVCaptureVideoPreviewLayer.m \
-AVComposition.m \
-AVCompositionTrack.m \
-AVCompositionTrackSegment.m \
-AVContentKeyReportGroup.m \
-AVContentKeyRequest.m \
-AVContentKeyResponse.m \
-AVContentKeySession.m \
-AVContentKeySpecifier.m \
-AVContinuityDevice.m \
-AVDelegatingPlaybackCoordinator.m \
-AVDepthData.m \
-AVExternalStorageDevice.m \
-AVExternalStorageDeviceDiscoverySession.m \
-AVExternalSyncDevice.m \
-AVFragmentedAsset.m \
-AVFragmentedAssetMinder.m \
-AVFragmentedAssetTrack.m \
-AVFragmentedMovie.m \
-AVFragmentedMovieMinder.m \
-AVFragmentedMovieTrack.m \
-AVMediaDataStorage.m \
-AVMediaSelection.m \
-AVMediaSelectionGroup.m \
-AVMediaSelectionOption.m \
-AVMetadataItemFilter.m \
-AVMetadataObject.m \
-AVMetadataBodyObject.m \
-AVMetadataCatBodyObject.m \
-AVMetadataDogBodyObject.m \
-AVMetadataFaceObject.m \
-AVMetadataHumanBodyObject.m \
-AVMetadataMachineReadableCodeObject.m \
-AVMetadataSalientObject.m \
-AVMovie.m \
-AVMovieTrack.m \
-AVMutableAudioMix.m \
-AVMutableAudioMixInputParameters.m \
-AVMutableComposition.m \
-AVMutableCompositionTrack.m \
-AVMutableMetadataItem.m \
-AVMutableMovie.m \
-AVMutableMovieTrack.m \
-AVMutableVideoComposition.m \
-AVMutableVideoCompositionInstruction.m \
-AVMutableVideoCompositionLayerInstruction.m \
-AVOutputSettingsAssistant.m \
-AVPersistableContentKeyRequest.m \
-AVPlaybackCoordinationMedium.m \
-AVPlaybackCoordinator.m \
-AVPlayerItemLegibleOutput.m \
-AVPlayerItemMetadataCollector.m \
-AVPlayerItemMetadataOutput.m \
-AVPlayerItemOutput.m \
-AVPlayerItemRenderedLegibleOutput.m \
-AVPlayerItemSampleBufferOutput.m \
-AVPlayerItemTrack.m \
-AVPlayerItemVideoOutput.m \
-AVPlayerLayer.m \
-AVPlayerLooper.m \
-AVPlayerPlaybackCoordinator.m \
-AVPlayerVideoOutput.m \
-AVPortraitEffectsMatte.m \
-AVQueuePlayer.m \
-AVRenderedCaptionImage.m \
-AVSampleBufferAudioRenderer.m \
-AVSampleBufferDisplayLayer.m \
-AVSampleBufferGenerator.m \
-AVSampleBufferGeneratorBatch.m \
-AVSampleBufferRenderSynchronizer.m \
-AVSampleBufferRequest.m \
-AVSampleBufferVideoRenderer.m \
-AVSemanticSegmentationMatte.m \
-AVSpeechSynthesisProviderAudioUnit.m \
-AVSpeechSynthesisVoice.m \
-AVSpeechSynthesizer.m \
-AVSpeechUtterance.m \
-AVSynchronizedLayer.m \
-AVTimedMetadataGroup.m \
-AVVideoComposition.m \
-AVVideoCompositionCoreAnimationTool.m \
-AVVideoCompositionInstruction.m \
-AVVideoCompositionLayerInstruction.m \
-AVVideoOutputSpecification.m \
-AVMutableTimedMetadataGroup.m \
 
 #
 # Makefiles
diff --git a/AVFoundation/GNUmakefile.preamble b/AVFoundation/GNUmakefile.preamble
index 987b375..abe107e 100644
--- a/AVFoundation/GNUmakefile.preamble
+++ b/AVFoundation/GNUmakefile.preamble
@@ -2,8 +2,13 @@
 # GNUmakefile.preamble
 #
 
-FFMPEG_CFLAGS := $(shell pkg-config --cflags libavformat libavcodec libavutil 2>/dev/null)
-FFMPEG_LIBS := $(shell pkg-config --libs libavformat libavcodec libavutil 2>/dev/null)
+FFMPEG_CFLAGS := $(shell pkg-config --cflags libavformat libavcodec libavutil libavdevice libswresample 2>/dev/null)
+FFMPEG_LIBS := $(shell pkg-config --libs libavformat libavcodec libavutil libavdevice libswresample 2>/dev/null)
+# Fallback if pkg-config returns empty (e.g. FreeBSD without .pc files)
+ifeq ($(strip $(FFMPEG_LIBS)),)
+FFMPEG_CFLAGS :=
+FFMPEG_LIBS   := -lavformat -lavcodec -lavutil -lavdevice -lswresample
+endif
 FLUIDSYNTH_CFLAGS := $(shell pkg-config --cflags fluidsynth 2>/dev/null)
 FLUIDSYNTH_LIBS := $(shell pkg-config --libs fluidsynth 2>/dev/null)
 
-- 
2.47.3

